matrix-macros.el

```diff
@@ -36,9 +162,189 @@ Is transformed to:
+                                           url-http-attempt-keepalives))
+         (callback (lambda (status &optional cbargs)
+                     (unwind-protect
+                         ;; This is called by `url-http-activate-callback' with the response buffer
+                         ;; as the current buffer.
+
+                         ;; Check for errors
+                         (pcase status
+                           ;; NOTE: This may need to be updated to correctly handle multiple errors
+                           (`(:error . ,_) (funcall error-body-fn
+						    :url url
+                                                    :cbargs cbargs
+                                                    :status status
+                                                    :error (plist-get status :error)))
+                           ((or 'nil
+                                `(:peer (:certificate . ,_))
+                                `(:redirect . ,_))
+                            (if (not url-http-end-of-headers)
+                                ;; HACK: It seems that the callback can be called with `nil' when
+                                ;; the connection fails before getting any headers, like:
+                                ;; url-http-end-of-document-sentinel(#<process matrix.org<5>>
+                                ;; "connection broken by remote peer\n"), in which case
+                                ;; `url-http-end-of-headers' is nil, so we need to call the error
+                                ;; fn.  Would like to structure this more cleanly.
+                                (funcall error-body-fn
+                                         :url url
+                                         :cbargs cbargs
+                                         :status status
+                                         :error (plist-get status :error))
+                              (let ((headers (buffer-substring (point) url-http-end-of-headers))
+                                    (data (if parser
+                                              (progn
+                                                (goto-char (1+ url-http-end-of-headers))
+                                                (funcall parser))
+                                            (buffer-substring (1+ url-http-end-of-headers) (point-max)))))
+                                (funcall success-body-fn
+                                         :cbargs cbargs
+                                         :status status
+                                         :headers headers
+                                         :data data))))
+                           (_ (error "Response status unrecognized; please report this error: %s" (pp-to-string status))))
+                       (when matrix-url-with-retrieve-async-timeout-timer
+                         (cancel-timer matrix-url-with-retrieve-async-timeout-timer))
+                       (unless (kill-buffer (current-buffer))
```

Comment:
Well, I dug through the url.el code, and the callback is called from inside the response buffer, so it will always be alive when this code runs.  It actually does not kill buffers on errors, either; it only marks them as dead internally, for killing whenever someone makes another request, from the `url-mm-callback` function, which is only called from `url-mm-url`.  So, we have to do it all ourselves.  That's the point of this function, after all, to cover over all these url.el issues.  :)